### PR TITLE
[Intel MKL] Adding support for FusedBatchNorm op with native format

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**

### CODE REVIEW

**Comment by gbaned** [comment]
@mahmoud-abuzaina  Can you please resolve conflicts? Thanks!

**Comment by mahmoud-abuzaina** [comment]
@gbaned I have resolved the conflicts. Thanks!

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for my delay! I have a few nits.

**Review by penpornk — CHANGES_REQUESTED** [review]
I have a few more comments that came up when I reviewed #43505.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 1]
```diff
@@ -1542,6 +1542,251 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
...
+        TF_RETURN_IF_ERROR(c->ReplaceDim(x, 1, channel_dim, &y));
+      }
+      c->set_output(0, y);
+      ShapeHandle vector_shape = c->Vector(channel_dim);
+      c->set_output(1, vector_shape);
+      c->set_output(2, vector_shape);
+      c->set_output(3, vector_shape);
+      c->set_output(4, vector_shape);
+      return Status::OK();
+    })
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: yes, done.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 2]
```diff
@@ -1542,6 +1542,251 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
...
+      // so that gradients can be performed when
+      // the op is in a symbolic condition.
+      if (is_training) {
+        c->set_output(3, c->Vector(0));
+        c->set_output(4, c->Vector(0));
+      } else {
+        c->set_output(3, c->Vector(channel_dim));
+        c->set_output(4, c->Vector(channel_dim));
+      }
+      return Status::OK();
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Fixed.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 3]
```diff
@@ -1542,6 +1542,251 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
...
+    })
+    .Doc(R"doc(
+MKL version of FusedBatchNormGrad operator that does not depend
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 4]
```diff
@@ -1542,6 +1542,251 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
...
+      c->set_output(1, vector_shape);
+      c->set_output(2, vector_shape);
+      c->set_output(3, vector_shape);
+      c->set_output(4, vector_shape);
+      return Status::OK();
+    })
+    .Doc(R"doc(
+MKL version of FusedBatchNorm operator that does not depend
+on layout propagation. Uses oneDNN APIs to perform fused
+batch normalization.
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 5]
```diff
@@ -1542,6 +1542,251 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
...
+    .Attr("U: {float}")
+    .Attr("epsilon: float = 0.0001")
+    .Attr("exponential_avg_factor: float = 1.0")
+    .Attr(GetConvnetDataFormatAttrString())
+    .Attr("num_side_inputs: int >= 0 = 0")
+    .Attr("activation_mode: string = \"Identity\"")
+    .Attr("is_training: bool = true")
+    .SetShapeFn(shape_inference::FusedBatchNormShape)
+    .Doc(R"doc(
+MKL version of FusedBatchNormEx operator that does not depend on layout propagation.
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 6]
```diff
@@ -986,7 +986,7 @@ class MklFusedBatchNormOp : public OpKernel {
       // Check if reorder is needed for src.
       const T* src_data = nullptr;
       std::shared_ptr<BatchNormFwdPd> bn_fwd_pd = bn_fwd->GetBatchNormFwdPd();
-      if (IS_SRC_REORDER_NEEDED(src_md, bn_fwd_pd, bn_fwd)) {
+      if (IS_SRC_REORDER_NEEDED(src_md, bn_fwd_pd, bn_fwd) && !native_format) {
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 7]
```diff
@@ -1381,7 +1389,8 @@ class MklFusedBatchNormGradOp : public OpKernel {
 
       // Check if diff_dst input needs to be reordered
       std::shared_ptr<BatchNormBwdPd> bn_bwd_pd = bn_bwd->GetBatchNormBwdPd();
-      if (IS_DIFF_DST_REORDER_NEEDED(diff_dst_md, bn_bwd_pd, bn_bwd)) {
+      if (IS_DIFF_DST_REORDER_NEEDED(diff_dst_md, bn_bwd_pd, bn_bwd) &&
+          !native_format) {
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc [hunk 8]
```diff
@@ -1390,7 +1399,7 @@ class MklFusedBatchNormGradOp : public OpKernel {
         diff_dst_data = static_cast<T*>(diff_dst.GetOpMem().get_data_handle());
       }
 
-      if (IS_SRC_REORDER_NEEDED(src_md, bn_bwd_pd, bn_bwd)) {
+      if (IS_SRC_REORDER_NEEDED(src_md, bn_bwd_pd, bn_bwd) && !native_format) {
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_fused_batch_norm_op.cc]: Done.

**Comment by mahmoud-abuzaina** [comment]
Thank you for the suggestions. I have addressed your comments.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Comment by gzmkl** [comment]
@penpornk thank you for the notification. I will work on the "cleanup" PR.

**Pull request merged**